dapp/contracts/Crowdsale.sol

```diff
@@ -149,15 +149,28 @@ contract Crowdsale is Ownable, BosonApproved {
             tokens = tokensLeft;
         }
 
+        // Note down the purchase.
         tokensLeft.sub(tokens);
 
         // Check if crowdsale is done.
         if (tokensLeft == 0) {
             nextState(State.Finished);
         }
 
+        // See if we need to send a refund. This comes when we sell less tokens
+        // than their value in Ether, which can happen when:
+        //  - we can't sell as many tokens as the buyer requested, because we
+        //    ran out of them
+        //  - the buyer send ether that cannot buy an entire token (ie. we
+        //    we refund the modulo of the price division)
+        uint256 refund = msg.value.sub(tokens.mul(price()));
+
         // Send tokens to buyer.
         require(token.transfer(buyer, tokens));
+        // Send refund to buyer, if necessary.
+        if (refund > 0) {
+            msg.sender.transfer(refund);
```

Comment:
Here's two strategies here. First one we don't actually care if the transfer back works because buying the tokens is more important, which I believe would be `msg.sender.send()` I think, and the second one is we revert the whole transaction in case of failure - `msg.sender.transfer()`. Other way would be to store the refund and allow the user to get it back at their own leisure when for example network unclogs

I'm, not sure which is better in terms of UX.

Reply:
I think the second/current option is the best - if we can't send back tokens, it's probably due to some extreme condition (or maliciousness from the user) - so we'd rather bail and not accept anything.

The first way is too profit-oriented for my taste. We have to be fair to our investors.

And the third way leads to a much more complex codebase for interacting with the contract, which I'd rather avoid.